relay-general/src/pii/event_paths.rs

```diff
@@ -0,0 +1,213 @@
+
+        // Walk through processing state in reverse order and build selector path off of that.
+        for substate in state.iter() {
+            if !substate.entered_anything() {
+                continue;
+            }
+
+            match substate.value_type() {
+                // $array.0.foo and $object.bar are not particularly good suggestions.
+                Some(ValueType::Object) | Some(ValueType::Array) => {}
+
+                // a.b.c.$string is not a good suggestion, so special case those.
+                ty @ Some(ValueType::String)
+                | ty @ Some(ValueType::Number)
+                | ty @ Some(ValueType::Boolean)
+                | ty @ Some(ValueType::DateTime) => {
+                    insert_path!(SelectorSpec::Path(vec![SelectorPathItem::Type(
+                        ty.unwrap()
+                    )]));
+                }
+
+                Some(ty) => {
+                    let mut path = path.clone();
+                    path.push(SelectorPathItem::Type(ty));
+                    path.reverse();
+                    if insert_path!(SelectorSpec::Path(path)) {
+                        // If we managed to generate $http.header.Authorization, we do not want to
+                        // generate request.headers.Authorization as well.
+                        return Ok(());
+                    }
+                }
+
+                None => {}
+            }
+
+            if let Some(key) = substate.path().key() {
+                path.push(SelectorPathItem::Key(key.to_owned()));
+            } else if substate.path().index().is_some() {
+                path.push(SelectorPathItem::Wildcard);
+            } else {
+                debug_assert!(substate.depth() == 0);
+                break;
+            }
+        }
+
+        if !path.is_empty() {
+            path.reverse();
+            insert_path!(SelectorSpec::Path(path));
+        }
+
+        Ok(())
+    }
+
+    fn process_pairlist<T: ProcessValue + AsPair>(
+        &mut self,
+        value: &mut PairList<T>,
+        _meta: &mut Meta,
+        state: &ProcessingState,
+    ) -> ProcessingResult {
+        process_pairlist(self, value, state)
+    }
+}
+
+/// Walk through the event and collect selectors that can be applied to it in a PII config. This
+/// function is used in the UI to provide auto-completion of selectors.
+///
+/// This generates a couple of duplicate suggestions such as `request.headers` and `$http.headers`
+/// in order to make it more likely that the user input starting with either prefix can be
+/// completed.
+///
+/// The main value in autocompletion is that we can complete `$http.headers.Authorization` as soon
+/// as the user types `Auth`.
+///
+/// XXX: This function should not have to take an event by value, we only do that
+/// due to restrictions on the Processor trait that we internally use to traverse the event.
+pub fn selectors_from_event(mut event: Annotated<Event>) -> BTreeSet<String> {
```

Comment:
In terms of the Rust interface, should this return a set (or vector) of `SelectorPath`s instead, and then only be stringified by the C-ABI?

Reply:
That's fair